### ISSUE TITLE
`select count(*) ...` queries on partial indexes should be constant

### ISSUE THREAD

**Issue by simerplaha** [alephium/explorer-backend#119]
# Overview

Count queries like the following on indexes that fully satisfy the where clause should be constant time.

### Query
```sql
select count(*) from block_headers where main_chain = true;
```

### Index
```sql
create index blocks_main_chain_idx on block_headers (main_chain) where main_chain = true;
```

# Issue

The query is not constant and is doing aggregate scan.

**Comment by simerplaha** [comment]
# Overview

When executing [list block](https://mainnet-backend.alephium.org/blocks) queries nearly 40% of the time is spent `counting` the rows in the `block_headers` table which is expensive. The performance worsens as the number of rows increases. So improving `count` query's performance is ideal.

![image](https://user-images.githubusercontent.com/1773953/159227995-c7d533de-b456-4792-bd32-71b1b3b8fe18.png)

# Improving performance

Unlike MySQL, PostgresSQL does not maintain any data structure that contains the count of rows for each table. Solutions implemented by others are: 

1. A dedicated `counter` table that maintains the row count value of each table via Postgres triggers.
2. Querying `pg_catalog.pg_class` table for approximating count.

The following are benchmarks for option 1. For the code see the commit above.

## Read performance improvement

The trigger improves the performance by a large margin because we only have to query the `user_row_counter` table which gets updated on each `INSERT`. 

![image](https://user-images.githubusercontent.com/1773953/159228691-edef7fd3-a5b9-4ecd-9220-1d8e56010f62.png)

## Cost of trigger on write performance

### Benchmarks when a single row is inserted
- `TriggerNone` - Trigger is disabled
- `TriggerForEachRow` - Trigger is enabled for each row insert
- `Trigger_None_ManualCountUpdate` - Trigger is disabled and a dedicated `counter_table`'s count is manually update with the transaction. 

![image](https://user-images.githubusercontent.com/1773953/159229225-92697c2d-6029-420f-942c-f4da7174483b.png)

**Conclusion:** Nearly 30-40% performance is lost inserting single row with a trigger. 

### Benchmarks when 200 rows are inserted as a transaction. 

`IndividualInserts` create a new `INSERT` query per row. 
```sql
BEGIN;
INSERT INTO TABLE values ("value1");
INSERT INTO TABLE values ("value2");
COMMIT;
```

![image](https://user-images.githubusercontent.com/1773953/159229246-312616f2-d6d6-4e42-ac8e-6fbc6b856c33.png)

**Conclusion:** Nearly 60% performance is lost inserting 200 rows with a trigger. 

### Benchmarks when 1000 rows are inserted. 

![image](https://user-images.githubusercontent.com/1773953/159229237-766de541-6134-4827-8f32-69fc2e6e98c3.png)

**Conclusion:** Over 90% performance is lost inserting 1000 rows with a trigger. 

# Conclusion

Triggers are painfully slow when inserting a larger transaction. But since we've restricted our batch inserts to 200 rows per query the performance loss might be bearable? Before saying yes, slow write queries also means that the database connection will be reserved by that write for some time, not being used to serve reads and we don't have as many cores. 

Unless I've missed something, within Postgres there is no perfect solution for improving `count` query's performance without paying for it in writes which would also cost a bit in reads by holding onto the connection. 

# Alternative solution

A cache in our Scala code that keeps the row count of necessary tables in-memory. Since we are using `Future` as return type for all queries we will be able to achieve this cache lock-free. 

Thoughts?

**Comment by polarker** [comment]
@simerplaha Thank you for the great investigation. It seems tricky indeed. Do you if this query would return results in constant time? `select count(*) from block_headers`

**Comment by polarker** [comment]
We have introduced Caffeine cache lately to `explorer-backend`, which support entry expiry. We could probably update the count of block pages every minute? It means there will be at most 1 query per minute. It's similar to your alternative solution. HDYT?

**Comment by tdroxler** [comment]
TIL about `triggers`
It's a nice timing you come with this, I'm adding transactions count for the FE statistics and I was wondering how well `select count(*) from transactions` was performing. I would have thought such "basic" query would have been fast.

The slowness of the insert with trigger is indeed painful, our syncing is already quite slow :/ 

For the transaction count I was going to do what Cheng proposed with the caffeine cache, but haven't thought further and didn't run any benchmark either, so I'll to have your opinion of this approach?

If we go the cache way, we still need to read from the db and can't only do it in-memory, because the explorer-backend is also launched in a read-only mode.

**Comment by polarker** [comment]
@tdroxler Much nicer communication than me ❤️

**Comment by simerplaha** [comment]
Totally! We should definitely store the row counts in Caffeine. It’s great as a cache and as a beverage :)

To build the cache during runtime we'd just piggy back off the sync process. We wouldn't need another `Thread` polling the database. 

Whenever we call `def insertAll(blocks: Seq[BlockEntity]): Future[Unit]` we'd do the following 
- All the `INSERT` queries return the count new rows added anyway. To build the cache we'd just sum it up into the cache to build the total count or dispatch `SELECT count(*) ...` after the end of the transaction. Whatever is easier, performs better & allowed by Postgres.
- If a `SELECT` query needs the row count of a table in cache (like the following `listMainChainSQL` function) while an `INSERT` is in progress, the cache (`fetchCountFromCache`) would return a `Promise` which would get completed after the `INSERT` is complete and the cache is updated. 

```scala
def listMainChainSQL(pagination: Pagination): Future[(Seq[BlockEntryLite], Int)] = {
  val blockEntries = run(listMainChainHeadersWithTxnNumberSQL(pagination))
  val count        = fetchCountFromCache(key = "mainChainCount_block_headers")
  blockEntries.zip(count)
}

```

This way we are not blocking any `Thread`, no polling & saving IO. Even if a request is waiting for a `Promise` to complete, that `Thread` is still working serving other queries.

In short, it's similar to [`ReentrantReadWriteLock`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/locks/ReentrantReadWriteLock.html) but async/lock-free and very minimal with `Future`.

**Comment by tdroxler** [comment]
> All the INSERT queries return the count of new rows added anyway. To build the cache we'd just sum it up into the cache to build the total count or dispatch SELECT count(*) ... after the end of the transaction. Whatever is easier, performs better & allowed by Postgres. 

The issue with that is for the `read-only` app, it won't get its cache updated.

Otherwise I like your other propositions

**Comment by simerplaha** [comment]
`read-only` would need the cache updated just ones on boot-up wouldn't it? The cache needs warm up stage on boot-up anyway, that runs the `count` queries initially on all the tables we want cached.

**Comment by polarker** [comment]
It's more efficient with your approach, but I have a few concerns:
1. As we are using a thread pool for `Future`, multiple threads would progress in parallel. I am afraid it's not possible to guarantee an order between `SELECT count()` and `INSERT`. `Future` is only lock-free when they are chained
2. It's tricky to return the number of rows added due to reorg and chain dependencies. We'd better to avoid working with reorg or chain dependencies as much as possible 😅

Using Caffeine with key expiry sounds much simpler, with the cost of one `SELECT count(*)` query per expiry interval.

**Comment by simerplaha** [comment]
>  I am afraid it's not possible to guarantee an order between `SELECT count()` and `INSERT`.

I wouldn’t have suggested it if it was not possible.

> reorg and chain dependencies.

Not sure what `reorg and chain dependencies` is.

I will just move forward with your preferred approach which you are right is much simpler. Please help me jot down how to progress with this so I understand how you want this done.

- Cache the count of tables we want cached by running `SELECT count(*) ...` and let Caffeine handle the expiration and re-population after `N` interval.
- What happens when there is time left to expiration (stale count in cache) and new `INSERT`s were added before the expiration?

**Comment by polarker** [comment]
I think @tdroxler is very familiar with it. @tdroxler Could you help?

**Comment by polarker** [comment]
> guarantee an order between SELECT count() and INSERT.

The tradeoff is: either you allow things to happen in parallel, or you sequence operations, right? I am willing to learn more and to be proven wrong as always.

**Comment by tdroxler** [comment]
> What happens when there is time left to expiration (stale count in cache) and new INSERTs were added before the expiration?

maybe the insert could invalidate the cache? This won't be applied to the read-only tho, but maybe we could afford having for 1-2 seconds some "non-fully-accurate" counts? AFAIK, the counts of blocks is used to show the pages number and the transaction count will be used for statistics, no need to be "real time" data IMO. WDYT?

(FYI I explained the reorg stuff at Simer on slack)

**Comment by polarker** [comment]
1. It's not critical at all to be precise in my opinion
2. the approximation is always less than the actual value, so it's fine for block query
